### PR TITLE
Allow non-Metabase users to unsubscribe from alerts

### PR DESCRIPTION
**Pull request by qwef**
For https://github.com/metabase/metabase/issues/29955

### Description

Publically exposed APIs that will subscribe/unsubscribe given the hash is correct. Also creates a job to periodically remove empty PulseChannels.

### How to verify

Submitting POST requests to the APIs, seeing the desired functionality. 

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/30597)

### CODE REVIEW

**Review by noahmoss — CHANGES_REQUESTED** [review]
Main request is changing the throttling to be on IP address; looks pretty good otherwise

File: src/metabase/api/session.clj [hunk 1]
```diff
@@ -313,4 +318,58 @@
         (log/error e (trs "Authentication endpoint error"))
         (throw e)))))
 
+;;; ----------------------------------------------------- Unsubscribe non-users from pulses -----------------------------------------------
+
+(def ^:private unsubscribe-throttler (throttle/make-throttler :email, :attempts-threshold 50))
+
+(defn generate-hash
```
[the remark this answers is not in the document]
Comment (qwef) [reply src/metabase/api/session.clj:325]: I plan to use this function in the next PR to send the different email

File: src/metabase/api/session.clj [hunk 2]
```diff
@@ -313,4 +318,58 @@
...
+   hash     :string}
+  (check-hash pulse-id email hash)
+  (api/let-404 [pulse-channel (t2/select-one PulseChannel :pulse_id pulse-id :channel_type "email")]
+               (let [emails       (get-in pulse-channel [:details :emails])
+                     given-email? #(= % email)]
+                 (if (some given-email? emails)
+                   (throw (ex-info (tru "Email for pulse-id already exists.")
+                                   {:type        type
+                                    :status-code 400}))
+                   (t2/update! PulseChannel (:id pulse-channel) (update-in pulse-channel [:details :emails] #(conj % email))))))
```
[the remark this answers is not in the document]
Comment (qwef) [reply src/metabase/api/session.clj]: O how does it know to put the argument before email, not after?

File: src/metabase/api/session.clj [hunk 3]
```diff
@@ -313,4 +318,57 @@
         (log/error e (trs "Authentication endpoint error"))
         (throw e)))))
 
+;;; ----------------------------------------------------- Unsubscribe non-users from pulses -----------------------------------------------
+
+(def ^:private unsubscribe-throttler (throttle/make-throttler :email, :attempts-threshold 50))
```
Comment (qwef) [inline src/metabase/api/session.clj:323]: What should I put as the argument here? I see the login throttler uses :username https://github.com/metabase/metabase/blob/master/src/metabase/api/session.clj#L80

File: src/metabase/api/session.clj (hunk 2 above)
[the remark this answers is not in the document]
Comment (noahmoss) [reply src/metabase/api/session.clj]: If you look at the [source](https://github.com/clojure/clojure/blob/clojure-1.10.1/src/clj/clojure/core.clj#L6172) for `update-in` it traverses the data structure with recursive logic, then calls `(apply f (get m k) args)` as the base case. So the value from the map being updated is the first argument, and then any remaining args are applied after it.

e.g.
```
> (update-in {:a {:b [1 2 3]}} [:a :b] conj 4 5)
{:a {:b [1 2 3 4 5]}}
```

File: src/metabase/api/session.clj (hunk 3 above)
Comment (noahmoss) [reply src/metabase/api/session.clj:323]: Here's the [description of that arg](https://github.com/metabase/throttle/blob/master/src/throttle/core.clj#L9-L10) from the throttle source. Seems like it's to convey to the frontend which field in a form the error should be applied to. Don't think it matters for this use case so you could just make it something like `:unsubscribe`

**Review by noahmoss — APPROVED** [review]

**Pull request merged**